testsuite/do_tests.sh

```diff
@@ -464,20 +467,26 @@ echo "Phase 7: Running PyNEST tests"
 echo "-----------------------------"
 
 if test "${PYTHON}"; then
-    PYNEST_TEST_DIR="${TEST_BASEDIR}/pytests/"
+    PYNEST_TEST_DIR="${TEST_BASEDIR}/pytests"
     XUNIT_NAME="07_pynesttests"
     
-    # Run all tests except those in the mpi subdirectory
+    # Run all tests except those in the mpi and non_concurrent subdirectories
     XUNIT_FILE="${REPORTDIR}/${XUNIT_NAME}.xml"
-    "${PYTEST}" --verbose --junit-xml="${XUNIT_FILE}" --numprocesses=auto \
-          --ignore="${PYNEST_TEST_DIR}/mpi" "${PYNEST_TEST_DIR}" 2>&1 | tee -a "${TEST_LOGFILE}" 
+    "${PYTHON}" -m pytest --verbose --timeout $TIME_LIMIT --junit-xml="${XUNIT_FILE}" --numprocesses=auto \
+          --ignore="${PYNEST_TEST_DIR}/mpi" --ignore="${PYNEST_TEST_DIR}/non_concurrent" \
+          "${PYNEST_TEST_DIR}" 2>&1 | tee -a "${TEST_LOGFILE}" 
+
+    # Run tests that cannot run concurrently
+    XUNIT_FILE="${REPORTDIR}/${XUNIT_NAME}_nc.xml"    
+    "${PYTHON}" -m pytest --verbose --timeout $TIME_LIMIT --junit-xml="${XUNIT_FILE}" \
+          "${PYNEST_TEST_DIR}/non_concurrent" 2>&1 | tee -a "${TEST_LOGFILE}" 
   
     # Run tests in the mpi subdirectories, grouped by number of processes
     if test "${HAVE_MPI}" = "true" -a "${MPI_LAUNCHER}" ; then
        for numproc in $(cd ${PYNEST_TEST_DIR}/mpi/; ls -d */ | tr -d '/'); do
            XUNIT_FILE="${REPORTDIR}/${XUNIT_NAME}_mpi_${numproc}.xml"
-           PYTEST_ARGS="--verbose --junit-xml=${XUNIT_FILE} ${PYNEST_TEST_DIR}/mpi/${numproc}"
-           $(sli -c "${numproc} (${PYTEST}) (${PYTEST_ARGS}) mpirun =only") 2>&1 | tee -a "${TEST_LOGFILE}"
+           PYTEST_ARGS="--verbose --timeout $TIME_LIMIT --junit-xml=${XUNIT_FILE} ${PYNEST_TEST_DIR}/mpi/${numproc}"
+           $(sli -c "${numproc} (${PYTHON} -m pytest) (${PYTEST_ARGS}) mpirun =only") 2>&1 | tee -a "${TEST_LOGFILE}"
```

Comment:
It's so spooky that line 489 actually works…
There is a severe mix of executables and parameters, and different types of quoting mixed with indirect evaluation, that should be cleaned up (not in this PR though). I would not have expected that a call to SLI is necessary to run the pytests, so I'd say this violates our principle of least surprise for me.

Please open an issue to refactor SLI out of the parallelization mechanics. I guess it's only doing config file parsing here to find information that could nowadays probably be found out by other means.

Reply:
@terhorstd I will create an issue. But we do not actually use SLI here to run NEST in parallel. We only use SLI to construct the correct invocation for running MPI jobs. It uses the user-adjusted `mpirun` command defined in `.nestrc`. This ensures that we always use the proper, system-specific way, even away from OpenMPI :).

Reply:
The NEST config file `.nestrc` is a SLI script that can only be read by executing it. I doubt this will change until the point when we disentangle SLI and NEST completely, which I'm not expecting anytime soon. But as this config file is really only needed for getting the MPI launcher command for running the testsuite, the situation is not so bad after all.

Reply:
Yes, I understand that SLI is not necessary to run pytests in parallel, which is exactly why I think this would be a positive first refactoring step. Requiring SLI just for parsing a string is what bothers me, especially since it already caused a major headache for exactly this wrong way of command line parsing. Just think about all characters that have a special meaning in SLI *or* bash *or* Python, be it a simple space in a path or so.
But probably you're right that we need to look at it in even broader scope. Maybe we find time to look into this at a smallish hackathon… – in a hut waaay out in the depth of nowhere.

Reply:
Please continue this discussion in #2154.